src/query/term_query/block_max_term_scorer.rs

```diff
@@ -0,0 +1,98 @@
+use crate::docset::{DocSet, SkipResult};
+use crate::query::{Explanation, Scorer};
+use crate::DocId;
+use crate::Score;
+
+use crate::fieldnorm::FieldNormReader;
+use crate::postings::Postings;
+use crate::postings::{BlockMaxPostings, BlockMaxSegmentPostings};
+use crate::query::bm25::BM25Weight;
+use crate::query::BlockMaxScorer;
+
+pub struct BlockMaxTermScorer {
+    postings: BlockMaxSegmentPostings,
+    fieldnorm_reader: FieldNormReader,
+    similarity_weight: BM25Weight,
+}
+
+impl BlockMaxTermScorer {
+    pub fn new(
+        postings: BlockMaxSegmentPostings,
+        fieldnorm_reader: FieldNormReader,
+        similarity_weight: BM25Weight,
+    ) -> Self {
+        Self {
+            postings,
+            fieldnorm_reader,
+            similarity_weight,
+        }
+    }
+}
+
+impl BlockMaxTermScorer {
+    fn _score(&self, fieldnorm_id: u8, term_freq: u32) -> Score {
+        self.similarity_weight.score(fieldnorm_id, term_freq)
+    }
+
+    pub fn term_freq(&self) -> u32 {
+        self.postings.term_freq()
+    }
+
+    pub fn fieldnorm_id(&self) -> u8 {
+        self.fieldnorm_reader.fieldnorm_id(self.doc())
+    }
+
+    pub fn explain(&self) -> Explanation {
+        let fieldnorm_id = self.fieldnorm_id();
+        let term_freq = self.term_freq();
+        self.similarity_weight.explain(fieldnorm_id, term_freq)
+    }
+}
+
+impl DocSet for BlockMaxTermScorer {
+    fn advance(&mut self) -> bool {
+        self.postings.advance()
+    }
+
+    fn skip_next(&mut self, target: DocId) -> SkipResult {
+        self.postings.skip_next(target)
+    }
+
+    fn doc(&self) -> DocId {
+        self.postings.doc()
+    }
+
+    fn size_hint(&self) -> u32 {
+        self.postings.size_hint()
+    }
+}
+
+impl Scorer for BlockMaxTermScorer {
+    fn score(&mut self) -> Score {
+        self._score(
+            self.fieldnorm_reader.fieldnorm_id(self.doc()),
+            self.postings.term_freq(),
+        )
+    }
+}
+
+impl BlockMaxScorer for BlockMaxTermScorer {
+    fn block_max_score(&mut self) -> Score {
+        self._score(
+            self.fieldnorm_reader
+                .fieldnorm_id(self.postings.block_max_doc()),
```

Comment:
this is inaccurate isn't it? The max score in the block is not necessarily reached for the posting list with the highest term freq in the block I believe.

Reply:
This deals with only one posting list, doesn't it? So there's only one block for one term, and so the only difference between them is the term frequency. The function would have to not be monotonic wrt. term frequency in order for this to be false. In fact, I believe the error I made was in the line below, where it should be `self.postings.block_max_term_freq()`. Or am I completely missing something here?

Reply:
The fieldnorm is the number of tokens per doc. It is not a constant over the block.

To get an upper bound of the score you would have to take the min fieldnorm, and the max termfreq over the block.

Reply:
Oh, yes, I see what you mean. The idea is to precompute that block max structure such that you don't store the max frequency but the frequency of the document with the max score.

Still, I realize that I made the following mistake: The block-max structure were supposed to be postings of this format: `(last document in the block, frequency of max-score posting within the block)`. This is so that you can find the block given your current document. But I also used that document ID as if it were the ID of the max-score document, which it is not.

In short, we need to store additional info. Off the top of my head, we can either (1) directly store scores instead of frequencies, or (2) store the document ID of the best posting as well.

In PISA, we do (1): We build the block max stuff per scoring function (bm25, query likelihood, etc.) and support storing it as raw uncompressed floats or compressed quantized integers.

The very easiest (but suboptimal) way of fixing this would probably be to have yet another posting list with document IDs of the max-score postings. But this immediately increases the size of the structure and unnecessarily stores block bound IDs twice. I'm not sure how difficult would be to generalize the segment postings to be able to store either floats or pairs of integers but I can take a look sometime this week. Reading quantized scores should be easy but that is additional complexity for the user if at any point you want to introduce other scoring functions.

Let me know what you think about all these and, as I said, I'll take a closer look soon.

Reply:
> The idea is to precompute that block max structure such that you don't store the max frequency but the frequency of the document with the max score.

That would work, but then this definition is contingent on the scoring function used... And the main benefit of storing the term freq for the maxscore and the fieldnorm for the doc with the max score disappears. We cannot use this with different scoring function.

One benefit I guess, could lie in the encoding opportunity it opens... but I don't think it should be part of the trait.

I suggest we simply encode the bm25 max score on a f16 straight in the skip list.,. at least for the moment.

Reply:
I guess you're right, it will be tied to a scoring function one way or another. We can start with floats and see how it goes. We can think of improving it later.